scopus-import/src/main/java/no/sikt/nva/scopus/conversion/AffiliationGenerator.java

```diff
@@ -1,121 +1,144 @@
+        return Optional.ofNullable(authorGroup.getAffiliation())
+                   .map(AffiliationType::getOrganization)
+                   .map(AffiliationGenerator::extractNames)
+                   .orElse(List.of());
     }
 
-    private static List<CristinOrganization> fetchCristinOrganizations(CristinConnection cristinConnection,
-                                                     List<String> organizationNames) {
-        return organizationNames.stream()
-                   .map(organization -> searchCristinOrganization(organization, cristinConnection))
-                   .filter(Optional::isPresent)
-                   .map(Optional::get)
-                   .toList();
+    private static String extractName(OrganizationTp org) {
+        return extractContentString(org.getContent());
     }
 
-    private static Optional<CristinOrganization> searchCristinOrganization(String organization,
-                                                                           CristinConnection cristinConnection) {
-        var searchResponse = cristinConnection.searchCristinOrganization(organization);
-        return searchResponse.isPresent() && hasSingleOrganization(searchResponse.get())
-                   ? Optional.ofNullable(getSingleOrganization(searchResponse.get()))
-                   : Optional.empty();
+    private static List<String> extractNames(List<OrganizationTp> organizationTps) {
+        return organizationTps.stream().map(AffiliationGenerator::extractName).toList();
     }
 
     private static CristinOrganization getSingleOrganization(SearchOrganizationResponse searchResponse) {
         logger.info("Fetched cristin organization from author affiliation labels: {}", searchResponse);
-        return searchResponse.hits().get(0);
+        return searchResponse.hits().getFirst();
     }
 
     private static boolean hasSingleOrganization(SearchOrganizationResponse searchOrganizationResponse) {
         return searchOrganizationResponse.size() == SINGLE_HIT;
     }
 
-    @NotNull
-    private static List<String> extractOrganizationNames(AuthorGroupTp authorGroup) {
-        return Optional.ofNullable(authorGroup.getAffiliation())
-                   .map(AffiliationType::getOrganization)
-                   .map(AffiliationGenerator::extractNames)
-                   .orElse(List.of());
+    private List<CorporationWithContributors> getCorporationsIds(List<AuthorGroupTp> authorGroupList) {
+        try (var customThreadPool = new ForkJoinPool(8)) {
+            var result =
+                customThreadPool.submit(() -> authorGroupList.parallelStream().map(this::retrieveCristinId))
+                    .join()
+                    .toList();
+            customThreadPool.shutdown();
+            return result;
+        } catch (Exception e) {
+            logger.warn("Parallel fetching of corporations id from PIA failed", e);
+            return authorGroupList.stream().map(this::retrieveCristinId).toList();
```

Comment:
I'm not sure how working with physical threads on Lambda works…there are some weird side effects of using lambda that we have to think about (see e.g. https://medium.com/p/multithreading-in-aws-lambda-part-1-performance-scaling-6e2bc3d7796f), we could try VirtualThreads (we're on Java 21…).

Reply:
I read some articles before I began this task. I think it as long as you call .join() to make sure the AWS Lambda was waiting for the result and shutdown() to prevent memory leak you were in safe?

Reply:
I will change to virtual threads to avoid some aws internals.